Why: Feature to deploy monit on target host and make the module work with newer versions of monit

prakritish opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Feature to deploy monit on target host and fix functionalities broken due to changed output in newer versions of monit.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
monit
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
[root@prakritish1 ansible]# ansible --version
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
[root@prakritish1 ansible]#
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
These changes adds new functionalities as well as solve a few bugs:
###### New Functionalities

1. Allows users to deploy/install **monit** on the target host by specifying either the _version_ or _URL_ for the **monit** binary. This does not use packaged distribution, but downloads directly from **monit** distribution website. The module identifies the platform and architecture of the target host and downloads the relevant binary and deploys them at the specified path.

1. Adds capabilities to specify the _**path**_ of the **monit** binary and it's _**config**_ file if they are not available in default path.

1. Added _service_ as an alias for _name_.

###### Bug Fix
With the last few versions of **monit**, the output of the binary was changed and this may have broken the module and none of the existing functionalities were working. So, made the changes to fix the broken functionalities.
<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

prakritish committed:
Why:
Feature to deploy monit on target host
*

This change addresses the need by:
The change allows user to specify a monit version or url of monit binary
to 'deploy' monit on target host. On specifying a version, the module
identifies the platform and architecture of the host and downloads the
required binary from mmonit.com and deploys it at the given path.

The other changes are to make this module work with newer versions of
monit. The module did not work with newer versions of monit because of
change in monit output.
*

prakritish committed:
Why:
Test Failed
*

This change addresses the need by:
Fixed the metadata and indentation issue as reported by ansibot
*

prakritish committed:
Why:
The test ansible-test sanity --test validate-modules failed with the
following errors:

lib/ansible/modules/monitoring/monit.py:0:0: E309 version_added for new
option (config) should be 2.4. Currently 0.0
lib/ansible/modules/monitoring/monit.py:0:0: E309 version_added for new
option (insecure) should be 2.4. Currently 0.0
lib/ansible/modules/monitoring/monit.py:0:0: E309 version_added for new
option (path) should be 2.4. Currently 0.0
lib/ansible/modules/monitoring/monit.py:0:0: E309 version_added for new
option (service) should be 2.4. Currently 0.0
lib/ansible/modules/monitoring/monit.py:0:0: E309 version_added for new
option (url) should be 2.4. Currently 0.0
lib/ansible/modules/monitoring/monit.py:0:0: E309 version_added for new
option (version) should be 2.4. Currently 0.0
*

*

bcoca commented on lib/ansible/modules/monitoring/monit.py:
```diff
@@ -23,26 +23,31 @@
 description:
      - Manage the state of a program monitored via I(Monit)
-version_added: "1.2"
+version_added: "1.3"
```
not sure why you are changing version here

prakritish replied:
Reverted back to "1.2" as suggested.

bcoca commented on lib/ansible/modules/monitoring/monit.py:
```diff
@@ -23,26 +23,31 @@
     required: true
     default: null
-    choices: [ "present", "started", "stopped", "restarted", "monitored", "unmonitored", "reloaded" ]
+    choices: [ "present", "started", "stopped", "restarted", "monitored", "unmonitored", "reloaded", "deployed" ]
```
I don't think installation should be covered by this modules, there are plenty of package management modules that are better suited for this, at worst a role that downloads and installs it

prakritish replied:
Removed "deployed" and the implementation as suggested.

prakritish committed:
Made the changes as requested by reviewer.

Reverted back the "version_added" to old value.
Removed the "deployed" state option.

prakritish commented:
@bcoca The requested changes are in place. Please review and let me know if any further changes are needed. 

**Playbook**
```
- hosts: monitHost
  become: yes
  become_method: sudo
  vars:
    version: 5.21.0
  tasks:
  - name: Start monit if not started
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      state: started
  - name: Check service {{ ohai_machinename }} present
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      service: "{{ ohai_machinename }}"
      state: "present"
  - name: Start service apache
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      name: "apache"
      state: "started"
  - name: Reload monit configuration
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      state: "reloaded"
  - name: Stop Service apache
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      service: "apache"
      state: "stopped"
  - name: Restart apache
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      service: "apache"
      state: "restarted"
  - name: Unmonitor Service apache
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      name: "apache"
      state: "unmonitored"
  - name: Monitor Service apache
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      service: "apache"
      state: "monitored"
  - name: Stop monit
    monit:
      path: "/opt/monit/bin/monit"
      config: "/etc/monitrc"
      state: "stopped"
```
**Output:**
```
[root@prakritish1 playbooks]# ansible-playbook monit.yml

PLAY [monitHost] ***************************************************************

TASK [setup] *******************************************************************
ok: [prakritish3]

TASK [Start monit if not started] **********************************************
changed: [prakritish3]

TASK [Check service prakritish3 present] ***********************
ok: [prakritish3]

TASK [Start service apache] ****************************************************
ok: [prakritish3]

TASK [Reload monit configuration] **********************************************
changed: [prakritish3]

TASK [Stop Service apache] *****************************************************
changed: [prakritish3]

TASK [Restart apache] **********************************************************
changed: [prakritish3]

TASK [Unmonitor Service apache] ************************************************
changed: [prakritish3]

TASK [Monitor Service apache] **************************************************
changed: [prakritish3]

TASK [Stop monit] **************************************************************
changed: [prakritish3]

PLAY RECAP *********************************************************************
prakritish3 : ok=10   changed=7    unreachable=0    failed=0

[root@prakritish1 playbooks]#
```

prakritish committed:
Why: Shebang was missing.
